crates/pathfinder/src/sync/stream.rs

```diff
@@ -0,0 +1,170 @@
+use futures::{Future, Stream, StreamExt, TryFutureExt, TryStream, TryStreamExt};
+use p2p::libp2p::PeerId;
+use p2p::PeerData;
+use tokio::sync::mpsc::Receiver;
+use tokio_stream::wrappers::ReceiverStream;
+
+use crate::sync::error::SyncError2;
+
+pub struct SyncReceiver<T> {
+    inner: Receiver<SyncResult<T>>,
+}
+pub type SyncResult<T> = Result<PeerData<T>, PeerData<SyncError2>>;
+
+pub trait ProcessStage {
+    type Input;
+    type Output;
+
+    fn map(
+        &mut self,
+        input: Self::Input,
+    ) -> impl Future<Output = Result<Self::Output, SyncError2>> + Send;
```

Comment:
It might be nicer to have this be a sync method instead. This is async to allow for using any streams that state might contain. But I can't think of anywhere that might be applicable.

Currently each implementation has to do a `spawn_blocking` which could instead be lifted into `SyncReceiver`.

Reply:
Yeah I think its probably just better 👍 Will do.

Reply:
Done